### ISSUE TITLE
NPE in ClientRequestResponseConverter (RxNetty 0.4.18)

### ISSUE THREAD

**Issue by dfjones** [ReactiveX/RxNetty#541]
In RxNetty 0.4.18 I started to see this NPE in some high load situations. I was able to reproduce it on RxNetty 0.4.16 as well. I think it is a race condition because I only see it when I crank up the concurrency in a load test that I have. I can't share this, unfortunately, but perhaps if this becomes tricky to debug I can whip up a shareable reproduction?

```
java.lang.NullPointerException
  at io.reactivex.netty.protocol.http.client.ClientRequestResponseConverter$ResponseState.sendOnError(ClientRequestResponseConverter.java:396)
  at io.reactivex.netty.protocol.http.client.ClientRequestResponseConverter$ResponseState.access$200(ClientRequestResponseConverter.java:350)
  at io.reactivex.netty.protocol.http.client.ClientRequestResponseConverter.exceptionCaught(ClientRequestResponseConverter.java:262)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:295)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:274)
  at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:266)
  at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:131)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:295)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:274)
  at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:266)
  at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:131)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:295)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:274)
  at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:266)
  at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireExceptionCaught(CombinedChannelDuplexHandler.java:423)
  at io.netty.channel.ChannelHandlerAdapter.exceptionCaught(ChannelHandlerAdapter.java:78)
  at io.netty.channel.CombinedChannelDuplexHandler$1.fireExceptionCaught(CombinedChannelDuplexHandler.java:144)
  at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:131)
  at io.netty.channel.CombinedChannelDuplexHandler.exceptionCaught(CombinedChannelDuplexHandler.java:230)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:295)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:274)
  at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:266)
  at io.netty.handler.logging.LoggingHandler.exceptionCaught(LoggingHandler.java:175)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:295)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:274)
  at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:266)
  at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:131)
  at io.reactivex.netty.pipeline.ssl.SslCompletionHandler.exceptionCaught(SslCompletionHandler.java:58)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:295)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:274)
  at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:266)
  at io.netty.handler.ssl.SslHandler.exceptionCaught(SslHandler.java:743)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:295)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:274)
  at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:266)
  at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:131)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:295)
  at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:274)
  at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:266)
  at io.reactivex.netty.pipeline.InternalReadTimeoutHandler.readTimedOut(InternalReadTimeoutHandler.java:178)
  at io.reactivex.netty.pipeline.InternalReadTimeoutHandler$ReadTimeoutTask.run(InternalReadTimeoutHandler.java:204)
  at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38)
  at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:120)
  at io.netty.util.concurrent.SingleThreadEventExecutor.safeExecute(SingleThreadEventExecutor.java:451)
  at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:418)
  at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:401)
  at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:877)
  at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
  at java.lang.Thread.run(Thread.java:745)
```

**Comment by dfjones** [comment]
I forgot to mention that these NPEs were seen after Netty read timeouts:

```
[rxnetty-nio-eventloop-3-2] WARN io.netty.channel.AbstractChannelHandlerContext - An exception 'java.lang.NullPointerException' [enable DEBUG level for full stacktrace] was thrown by a user handler's exceptionCaught() method while handling the following exception:
io.netty.handler.timeout.ReadTimeoutException
```

**Comment by NiteshKant** [comment]
@dfjones apologies for the delay in responding.

Looking at the code, it is possible for `connInputObsrvr` to be `null`. It looks like you are using SSL, it's just a wild guess but can you try and repro this without SSL?

**Comment by dfjones** [comment]
@NiteshKant No worries! Thanks for the response.

Your guess was good, I turned off SSL and could not reproduce the issue.

**Comment by NiteshKant** [comment]
😄  I should have bought a lottery yesterday! (Well, it was an educated guess!!)

I think SSL handshake writes starts the read timeout tick and in some cases, timeout triggers before the request is written which causes the `connInputObserver` to be null. So, I think it makes sense to add a null check there for connection close. 
Do you want to send a PR?

**Comment by dfjones** [comment]
@NiteshKant Sure! PR created: https://github.com/ReactiveX/RxNetty/pull/544

Also, from my testing I had to guard against null for `connection` as well. With this PR my test case runs without error.

**Comment by NiteshKant** [comment]
Fixed via #545
